### PR TITLE
Introduce pthread_mutex APIs.

### PR DESCRIPTION
**Pull request by y4vor**
- Add support for pthread_mutex_init, pthread_mutex_destroy, pthread_mutex_lock, pthread_mutex_unlock and pthread_mutex_trylock.

b/302335657

Change-Id: I8774ddb8ad466b7342e045402b170a719f2e1a81

### CODE REVIEW

File: third_party/abseil-cpp/absl/base/internal/sysinfo.cc [hunk 1]
```diff
@@ -418,7 +418,11 @@ pid_t GetTID() {
 
 // Fallback implementation of GetTID using pthread_getspecific.
 ABSL_CONST_INIT static once_flag tid_once;
+#if defined(STARBOARD)
+static pthread_key_t tid_key;
```
[the remark this answers is not in the document]
Comment (y4vor) [reply third_party/abseil-cpp/absl/base/internal/sysinfo.cc:427]: Yes, the build fails as the compiler can't create a constant initializer for our type.
Added comment.

File: starboard/shared/win32/posix_emu/include/pthread.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef STARBOARD_SHARED_WIN32_POSIX_EMU_INCLUDE_PTHREAD_H_
+#define STARBOARD_SHARED_WIN32_POSIX_EMU_INCLUDE_PTHREAD_H_
+
+#include <windows.h>
+
+#ifdef __cplusplus
+#define PTHREAD_MUTEX_INITIALIZER \
```
[the remark this answers is not in the document]
Comment (y4vor) [reply starboard/shared/win32/posix_emu/include/pthread.h]: yes it is used in the compliance tests.
Updated the code to map it to SRWLOCK_INIT which is more accurate. It does the same, but it is cleaner.

File: third_party/musl/src/starboard/pthread/pthread.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#define THIRD_PARTY_MUSL_SRC_STARBOARD_PTHREAD_H_
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef __cplusplus
+#define PTHREAD_MUTEX_INITIALIZER \
```
[the remark this answers is not in the document]
Comment (y4vor) [reply third_party/musl/src/starboard/pthread/pthread.h:25]: Both compliance tests and third party code use it.

File: third_party/musl/include/sys/types.h [hunk 4]
```diff
@@ -30,8 +30,8 @@ extern "C" {
 #define __NEED_suseconds_t
 #define __NEED_blksize_t
 
+#if !defined(STARBOARD)
 #define __NEED_pthread_t
-#define __NEED_pthread_attr_t
```
[the remark this answers is not in the document]
Comment (y4vor) [reply third_party/musl/include/sys/types.h:34]: yes

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+};
+
+const int kLoops = 10000;
+
+void* EntryPoint(void* parameter) {
+  TestContext* context = static_cast<TestContext*>(parameter);
+
+  for (int i = 0; i < kLoops; ++i) {
+    pthread_mutex_lock(&context->mutex);
+    int k = context->count;
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc:39]: Is this preferred over `context->count++` for some reason?

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+TEST(PosixMutexAcquireTest, SunnyDayContended) {
+  TestContext context;
+  EXPECT_EQ(pthread_mutex_init(&context.mutex, NULL), 0);
+  const int kThreads = 4;
+  SbThread threads[kThreads];
+  for (int i = 0; i < kThreads; ++i) {
+    threads[i] = SbThreadCreate(0, kSbThreadNoPriority, kSbThreadNoAffinity,
+                                true, NULL, EntryPoint, &context);
+  }
+
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc:59]: Shouldn't we check here that `context.count == kThreads*kLoops`?

File: third_party/abseil-cpp/absl/base/internal/sysinfo.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewsavage1) [reply third_party/abseil-cpp/absl/base/internal/sysinfo.cc:427]: Should we instead change the definition of ABSL_CONST_INIT when STARBOARD is defined? I.e. do we expect to run into a similar problem for constant initialization of other types in this file or not?

[the remark this answers is not in the document]
Comment (y4vor) [reply third_party/abseil-cpp/absl/base/internal/sysinfo.cc:427]: That would affect more code outside of threading. This a very localized issue.
There might be other types yes.

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc (hunk 5 above)
Comment (y4vor) [reply starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc:39]: no. Let me update it.  The bulk of the test is coming from the existing NPLB test.

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc (hunk 6 above)
Comment (y4vor) [reply starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc:59]: No. The increments are happening on other threads. There is no guarantee all of them happened unless we block and wait on some condition.

File: starboard/shared/modular/BUILD.gn [hunk 7]
```diff
@@ -18,11 +18,16 @@ source_set("posix_wrappers") {
...
     "posix_time_wrappers.cc",
     "posix_time_wrappers.h",
   ]
 
   configs += [ "//starboard/build/config:starboard_implementation" ]
 
-  deps = [ "//starboard:starboard_headers_only" ]
+  deps = [
+    "//starboard:starboard_headers_only",
+    "//starboard/common",
```
[the remark this answers is not in the document]
Comment (y4vor) [reply starboard/shared/modular/BUILD.gn:31]: Don't need it anymore. Removed.

File: starboard/nplb/posix_compliance/posix_mutex_acquire_try_test.cc [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+TEST(PosixMutexAcquireTest, SunnyDayAutoInit) {
+  pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+  EXPECT_EQ(pthread_mutex_trylock(&mutex), 0);
+  EXPECT_EQ(pthread_mutex_unlock(&mutex), 0);
+  EXPECT_EQ(pthread_mutex_destroy(&mutex), 0);
+}
+
+TEST(PosixMutexAcquireTryTest, RainyDayReentrant) {
```
[the remark this answers is not in the document]
Comment (y4vor) [reply starboard/nplb/posix_compliance/posix_mutex_acquire_try_test.cc:58]: I'm following the existing NPLB test cases.
We are planning on replacing them with a third_party posix test suite.
If that gets delayed too much we'll feef up these tests.

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  for (int i = 0; i < kLoops; ++i) {
+    for (int j = 0; j < kThreads; ++j) {
+      EXPECT_EQ(pthread_mutex_lock(&context.mutex), 0);
+      int k = context.count;
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc:61]: Let's do the same cleanup here of `context.count--`

**Review by andrewsavage1 — APPROVED** [review]

File: starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc (hunk 9 above)
Comment (y4vor) [reply starboard/nplb/posix_compliance/posix_mutex_acquire_test.cc]: done

**Pull request merged**